app/services/authentication/authenticator.rb

```diff
@@ -26,24 +26,46 @@ def self.call(*args)
 
     def call
       identity = Identity.build_from_omniauth(provider)
-
       return current_user if current_user_identity_exists?
 
-      user = proper_user(identity)
-      user = if user.nil?
-               build_user
-             else
-               update_user(user)
-             end
+      # These variables need to be set outside of the scope of the
+      # transaction in order to be used after the transaction is completed.
+      log_to_datadog = false
+      id_provider, authed_user = nil
+
+      ActiveRecord::Base.transaction do
+        user = proper_user(identity)
+        user = if user.nil?
+                 find_or_create_user!
+               else
+                 update_user(user)
+               end
+
+        identity.user = user if identity.user_id.blank?
+        new_identity = identity.new_record?
+        successful_save = identity.save!
 
-      save_identity(identity, user)
+        log_to_datadog = new_identity && successful_save
+        id_provider = identity.provider
 
-      user.skip_confirmation!
+        user.skip_confirmation!
+
+        flag_spam_user(user) if account_less_than_a_week_old?(user, identity)
+
+        user.save!
+        authed_user = user
+      end
 
-      flag_spam_user(user) if account_less_than_a_week_old?(user, identity)
+      if log_to_datadog
+        # Notify DataDog if a new identity was successfully created.
+        DatadogStatsClient.increment("identity.created", tags: [provider: id_provider])
+      end
 
-      user.save!
-      user
+      # Return the successfully-authed used from the transaction.
+      authed_user
+    rescue StandardError => e
+      # Notify DataDog if something goes wrong in the transaction.
+      DatadogStatsClient.increment("identity.errors", tags: ["error:#{e.class}"], message: e.message)
```

Comment:
I think we need to re-raise the exception after logging to DD, otherwise the user will never now what went wrong, correct? This is also why the test fails

Reply:
I would also add a test in the authenticator for that

Reply:
Done!